mathjax3-ts/output/chtml/FontData.ts

```diff
@@ -0,0 +1,328 @@
+    protected static defaultChars: CharMapMap = {};
+
+    /*
+     * The default variants for the fixed size stretchy delimiters
+     */
+    protected static defaultSizeVariants: string[] = [];
+
+    /*
+     * The actual variant, delimiter, and size information for this font
+     */
+    protected variant: VariantMap = {};
+    protected delimiters: DelimiterMap = {};
+    protected sizeVariants: string[];
+
+    /*
+     * The actual font parameters for this font
+     */
+    public params: FontParameters;
+
+    /*
+     * Copies the data from the defaults to the instance
+     *
+     * @constructor
+     */
+    constructor() {
+        let CLASS = (this.constructor as typeof FontData);
+        this.params = Object.assign({}, CLASS.defaultParams);
+        this.sizeVariants = CLASS.defaultSizeVariants.slice(0);
+        this.createVariants(CLASS.defaultVariants);
+        this.defineDelimiters(CLASS.defaultDelimiters);
+        for (const name of Object.keys(CLASS.defaultChars)) {
+            this.defineChars(name, CLASS.defaultChars[name]);
+        }
+    }
+
+    /*
+     * Creates the data structure for a variant (an object with prototype chain
+     *   that includes a copy of the linked variant, and then the inherit variant chain.
+     *   The linked copy is updated automatically when the link variant is modified.
+     *   (The idea is to be able to have something like bold-italic inherit from both
+     *   bold and intalic by having the prototype chain include a copy of bold plus
+     *   the full italic chain.  So if something is not defined explicitly in bold-italic,
+     *   it defaults first to a bold version, than an italic version, then the normal
+     *   version, which is in the italic prototype chain.)
+     *
+     * @param{string} name     The new variant to create
+     * @param{string} inherit  The variant to use if a character is not in this one
+     * @param{string} link     A variant to search before the inherit one (but only
+     *                           its top-level object).
+     */
+    public createVariant(name: string, inherit: string = null, link: string = null) {
+        let variant = {
+            linked: [] as CharMap[],
+            chars: (inherit ? Object.create(this.variant[inherit].chars) : {}) as CharMap
+        };
+        if (link && this.variant[link]) {
+            Object.assign(variant.chars, this.variant[link].chars);
+            this.variant[link].linked.push(variant.chars);
+            variant.chars = Object.create(variant.chars);
+        }
+        this.variant[name] = variant;
+    }
+
+    /*
+     * Create a collection of variants
+     *
+     * @param{string[][]} variants  Array of [name, inherit?, link?] values for
+     *                              the variants to define
+     */
+    public createVariants(variants: string[][]) {
+        for (const variant of variants) {
+            this.createVariant.apply(this, variant);
```

Comment:
Is there a reason why you are using .apply here?

Reply:
Because variant is an array that contains the `name`, `inherit`, and `link` values for the `createVariant()` call, so they need to be broken out.  I suppose you could do

```
this.createVariant(variant[0], variant[1], variant[2]);
``` 
or
```
this.createVariant(...variant);
```
if you prefer.  (I admit that I don't always think of the spread operator; It is hard to unlearn past paradigms.)

Reply:
It turns out Typescript doesn't like the `this.createVariant(...variant)` form, so I used the other more verbose one.